### DISCUSSION TITLE
dataclass only fields not set when reading from the database

### DISCUSSION THREAD

**Discussion by lucasgadams** [sqlalchemy/sqlalchemy#9928] [Usage Questions]
### Describe the bug

When mixing `mapped_column` fields with native dataclass `fields` with `default_factory`, the dataclass-only fields are not set when reading an object back from the database. This makes using the dataclass fields for anything other than dynamic object *creation* difficult. When creating an object the `default_factory` hook is correctly called and the dataclass fields will be created correctly. But if you then read the object back from the database, it seems that the dataclass only fields are not initialized. I would expect that some sort of initialization of these fields should happen on read. 

In my specific example I am storing a certain state variable as a dataclass only field. This state is just used in python to keep track of certain events. I set the field up with `dataclass_field(default_factory=list, init=False)`. **I would expect**: when reading the model back from the database the `default_factory` hook is still called, and the state variable should then exist (initialized to a new empty list). Instead the attribute is completely missing. I also tried using `__post_init__` but that did not work either. I tried a few different combos of `repr`, `init`, `kw_only`, but that did not help. I have tried this across sqllite and asyncpg. 

My current workaround is just to create property for the field that will either create or return the state variable, and not use the dataclass functionality. 

### SQLAlchemy Version in Use

2.0.15

### DBAPI (i.e. the database driver)

pysqlite

### Database Vendor and Major Version

SQLite

### Python Version

3.11.2

### Operating system

OSX

### To Reproduce

```python
import logging
import uuid
from dataclasses import field as dataclass_field

from sqlalchemy import create_engine, select
from sqlalchemy.orm import (
    DeclarativeBase,
    Mapped,
    MappedAsDataclass,
    Session,
    mapped_column,
)

logger = logging.getLogger(__name__)
logging.basicConfig(level=logging.INFO)

engine = create_engine("sqlite+pysqlite:///:memory:", echo=True)


class Base(MappedAsDataclass, DeclarativeBase):
    pass


class Model(Base):
    __tablename__ = "models_test"
    id: Mapped[uuid.UUID] = mapped_column(
        primary_key=True, default_factory=uuid.uuid4, init=False
    )
    name: Mapped[str | None] = mapped_column(default=None)

    dataclass_state: list[int] = dataclass_field(default_factory=list, init=False)


def main():
    with engine.begin() as conn:
        Base.metadata.drop_all(conn)

    with engine.begin() as conn:
        Base.metadata.create_all(conn)

    with Session(engine, expire_on_commit=False) as session:
        model = Model(name="test")
        logger.info(f"*** MODEL: {model} ***")
        model.dataclass_state.append(1)
        logger.info(f"*** MODEL: {model} ***")
        session.add(model)
        session.commit()
        logger.info(f"*** MODEL: {model} ***")
        logger.info(f"*** DC attribute before read: {model.dataclass_state} ***")
        model_id = model.id

    with Session(engine, expire_on_commit=False) as session2:
        model2 = session2.execute(
            select(Model).filter(Model.id == model_id)
        ).scalar_one()
        logger.info(f"*** SA attributes after read: {model2.id}, {model2.name} ***")
        logger.info(f"*** DC attribute after read: {model2.dataclass_state} ***")


if __name__ == "__main__":
    main()
```

### Error

```
2023-06-10 10:23:20,480 INFO sqlalchemy.engine.Engine BEGIN (implicit)
INFO:sqlalchemy.engine.Engine:BEGIN (implicit)
2023-06-10 10:23:20,480 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("models_test")
INFO:sqlalchemy.engine.Engine:PRAGMA main.table_info("models_test")
2023-06-10 10:23:20,480 INFO sqlalchemy.engine.Engine [raw sql] ()
INFO:sqlalchemy.engine.Engine:[raw sql] ()
2023-06-10 10:23:20,480 INFO sqlalchemy.engine.Engine PRAGMA temp.table_info("models_test")
INFO:sqlalchemy.engine.Engine:PRAGMA temp.table_info("models_test")
2023-06-10 10:23:20,480 INFO sqlalchemy.engine.Engine [raw sql] ()
INFO:sqlalchemy.engine.Engine:[raw sql] ()
2023-06-10 10:23:20,480 INFO sqlalchemy.engine.Engine COMMIT
INFO:sqlalchemy.engine.Engine:COMMIT
2023-06-10 10:23:20,481 INFO sqlalchemy.engine.Engine BEGIN (implicit)
INFO:sqlalchemy.engine.Engine:BEGIN (implicit)
2023-06-10 10:23:20,481 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("models_test")
INFO:sqlalchemy.engine.Engine:PRAGMA main.table_info("models_test")
2023-06-10 10:23:20,481 INFO sqlalchemy.engine.Engine [raw sql] ()
INFO:sqlalchemy.engine.Engine:[raw sql] ()
2023-06-10 10:23:20,481 INFO sqlalchemy.engine.Engine PRAGMA temp.table_info("models_test")
INFO:sqlalchemy.engine.Engine:PRAGMA temp.table_info("models_test")
2023-06-10 10:23:20,481 INFO sqlalchemy.engine.Engine [raw sql] ()
INFO:sqlalchemy.engine.Engine:[raw sql] ()
2023-06-10 10:23:20,481 INFO sqlalchemy.engine.Engine
CREATE TABLE models_test (
	id CHAR(32) NOT NULL,
	name VARCHAR,
	PRIMARY KEY (id)
)


INFO:sqlalchemy.engine.Engine:
CREATE TABLE models_test (
	id CHAR(32) NOT NULL,
	name VARCHAR,
	PRIMARY KEY (id)
)


2023-06-10 10:23:20,481 INFO sqlalchemy.engine.Engine [no key 0.00003s] ()
INFO:sqlalchemy.engine.Engine:[no key 0.00003s] ()
2023-06-10 10:23:20,481 INFO sqlalchemy.engine.Engine COMMIT
INFO:sqlalchemy.engine.Engine:COMMIT
INFO:__main__:*** MODEL: Model(id=UUID('2dbbf905-4947-4dde-8cdf-aaa89e0fe0ca'), name='test', dataclass_state=[]) ***
INFO:__main__:*** MODEL: Model(id=UUID('2dbbf905-4947-4dde-8cdf-aaa89e0fe0ca'), name='test', dataclass_state=[1]) ***
2023-06-10 10:23:20,481 INFO sqlalchemy.engine.Engine BEGIN (implicit)
INFO:sqlalchemy.engine.Engine:BEGIN (implicit)
2023-06-10 10:23:20,482 INFO sqlalchemy.engine.Engine INSERT INTO models_test (id, name) VALUES (?, ?)
INFO:sqlalchemy.engine.Engine:INSERT INTO models_test (id, name) VALUES (?, ?)
2023-06-10 10:23:20,482 INFO sqlalchemy.engine.Engine [generated in 0.00006s] ('2dbbf90549474dde8cdfaaa89e0fe0ca', 'test')
INFO:sqlalchemy.engine.Engine:[generated in 0.00006s] ('2dbbf90549474dde8cdfaaa89e0fe0ca', 'test')
2023-06-10 10:23:20,482 INFO sqlalchemy.engine.Engine COMMIT
INFO:sqlalchemy.engine.Engine:COMMIT
INFO:__main__:*** MODEL: Model(id=UUID('2dbbf905-4947-4dde-8cdf-aaa89e0fe0ca'), name='test', dataclass_state=[1]) ***
INFO:__main__:*** DC attribute before read: [1] ***
2023-06-10 10:23:20,482 INFO sqlalchemy.engine.Engine BEGIN (implicit)
INFO:sqlalchemy.engine.Engine:BEGIN (implicit)
2023-06-10 10:23:20,483 INFO sqlalchemy.engine.Engine SELECT models_test.id, models_test.name
FROM models_test
WHERE models_test.id = ?
INFO:sqlalchemy.engine.Engine:SELECT models_test.id, models_test.name
FROM models_test
WHERE models_test.id = ?
2023-06-10 10:23:20,483 INFO sqlalchemy.engine.Engine [generated in 0.00005s] ('2dbbf90549474dde8cdfaaa89e0fe0ca',)
INFO:sqlalchemy.engine.Engine:[generated in 0.00005s] ('2dbbf90549474dde8cdfaaa89e0fe0ca',)
INFO:__main__:*** SA attributes after read: 2dbbf905-4947-4dde-8cdf-aaa89e0fe0ca, test ***
2023-06-10 10:23:20,483 INFO sqlalchemy.engine.Engine ROLLBACK
INFO:sqlalchemy.engine.Engine:ROLLBACK
Traceback (most recent call last):
  File "/Users/lucasadams/Repos/materia-services/src/test_sa.py", line 61, in <module>
    main()
  File "/Users/lucasadams/Repos/materia-services/src/test_sa.py", line 57, in main
    logger.info(f"*** DC attribute after read: {model2.dataclass_state} ***")
                                                ^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'Model' object has no attribute 'dataclass_state'
```

**Comment by zzzeek** [comment]
The SQLAlchemy ORM does not call the `__init__` method when it loads objects.  current docs seem to be lacking a clear statement of this behavior, and given the dataclass feature we should definitely add more notes / warnings that `__init__` is not called on load.

for a description of the behavior you can refer to the [load()](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.InstanceEvents.load) event.   any non-ORM initialization that you want to occur upon load you would need to implement using this event hook.

**Reply by zzzeek** [reply]
we definitely need to refresh docs here.  there was a legacy section that is still linked from the load doc which was removed.

**Reply by lucasgadams** [reply]
I believe this will cause a lot of headache as the supposed dataclass functionality makes it seem like these mapped classes should behave in a way that they dont. For example:
- now that the SA models implement the dataclass transform, functions like `dataclasses.is_dataclass` return True. At least some frameworks interpret that to mean functions like `dataclasses.asdict` should work, but in the example I gave that will fail if the model is read from the db rather than instantiated directly.
- specifically, FastAPI/pydantic has support for dataclasses as well. They also have their own dataclass wrapper. Two things can happen here if you arent careful:
  - In standard pydantic you can use a dataclass model as a field inside of a pydantic model. Under the hood pydantic converts this to their dataclass wrapper. This seems to happen automatically if you use a SA dataclass model. So now you have a pydantic dataclass wrapping a SA dataclass. And pydantic then tries to validate the field, but the declared attributes of the dataclass, `dataclass_state` in my example, are missing after read so this fails. 
  - In FastAPI, you can use the `response_model` with a pydantic model with `orm_mode = True`, and should be able to just return a SA model from your endpoint function and it gets parsed into a pydantic model through attribute lookup. This works, *except* if the SA model is dataclass enabled and you are using a mix of SA and dataclass fields. In *that* case, FastAPI detects the incoming object as a `dataclass`, and tries to do things like call `asdict`. If it is *not* a SA-dataclass (just a standard SA model), the FastAPI correctly uses the `from_orm` function. If it *is a SA dataclass*, FastAPI thinks it should just use the "safer" dataclass methods to convert the object. But this will fail if you have read the object from the DB and have un-initialized dataclass fields. 
  
- you can get around these issues by explicitly converting things yourself where needed, using  `orm_mode=True`, `from_orm` and `parse_obj_as` from pydantic. 
  
Making the new SA dataclasses work well with pydantic/FastAPI has been quite difficult.

**Reply by lucasgadams** [reply]
And you absolutely need to hide these dataclass fields with `repr=False` or else the repr function will be completely broken (one of the main purported benefits of using SA-dataclasses in the first place...)

**Reply by lucasgadams** [reply]
Here is a place in FastAPI where this happens https://github.com/tiangolo/fastapi/blob/master/fastapi/routing.py#L105. FYI @tiangolo

**Reply by zzzeek** [reply]
as I've been telling others, @tiangolo  has written [SQLModel](https://sqlmodel.tiangolo.com/) which IMO should be where all the "pydantic integration" use cases should start first.  SQLAlchemy does not offer pydantic integration out of the box.

**Reply by lucasgadams** [reply]
Yeah i actually had been using that and waiting for sqlalchemy 2.0. Just spent like 2 weeks migrating to sqlalchemy. My understanding was that all of these things should become easier now with 2.0. The SA docs give a nod to SQLModel as inspiration for 2.0, and while I love with @tiangolo has done, SQLModel is not really being worked on much, never made a 1.0 release, is not compatible with SA 2.0, and has some issues. Again, I had thought that SA 2.0 would really be bridging the gap more, and SQLModel was just to make some things work in the meantime. Honestly FastAPI is so big, I would really hope to see better support from SA.

**Reply by CaselIT** [reply]
it seems that pydantic version 2 will be more compatible in general, see for example https://github.com/sqlalchemy/sqlalchemy/issues/9905 that seems to work fine with pydantic 2

Regarding this, I don't think it's feasible to change the current behavior at lest on 2.0, since it would be a somewhat large backward incompatible change. Also from a performance standpoint it would be less than ideal.

For sure better documentation can be added.

**Comment by CaselIT** [comment]
I don't think mixing mapped and unmapped dataclass fieds is a good approach. It probably should be documented
I think what you should instead use is a `cached_property` for your `dataclass_state`

**Reply by lucasgadams** [reply]
There is a whole section in the documentation describing this functionality https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#using-non-mapped-dataclass-fields

**Reply by zzzeek** [reply]
in #9930 we will indicate that these fields are not automatically populated on load.   it would be incorrect for SQLAlchemy to automatically apply "defaults" to an object being loaded from the database, since objects being loaded have existing state.  this state is undefined for fields that have no persistence definition, therefore the abovementioned `load()` event (or special methods on objects invoked by a more generic `load()` event) must be used in order to define the loading from persistence behavior for these fields.

**Reply by CaselIT** [reply]
> There is a whole section in the documentation describing this functionality https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#using-non-mapped-dataclass-fields

interesting. In hindsight maybe it wasn't the best decision. Opinions @zzzeek ?

for your use case `cached_property` is probably a better option though

**Reply by zzzeek** [reply]
we had users that were using plain dataclass fields, in my recollection, and I certainly dont want to disallow it.   easy enough to use load() / refresh() hooks to define their behavior, or if we get very fancy we could add new methods to declarative `__on_load__()` `__on_refresh__()` etc though that might get out of hand as there are I think three or four varieties of event here

**Reply by CaselIT** [reply]
I'll try to write some clarification notes there then
